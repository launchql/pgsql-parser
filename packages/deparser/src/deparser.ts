import { Node } from '@pgsql/types';
import { SqlFormatter } from './utils/sql-formatter';
import { ConsolidatedVisitor } from './visitors/consolidated-visitor';
import { DeparserContext } from './visitors/base';

export interface DeparserOptions {
  newline?: string;
  tab?: string;
}

export class Deparser {
  private formatter: SqlFormatter;
  private consolidatedVisitor: ConsolidatedVisitor;
  private tree: Node[];

  constructor(tree: Node | Node[], opts: DeparserOptions = {}) {
    this.formatter = new SqlFormatter(opts.newline, opts.tab);
    this.consolidatedVisitor = new ConsolidatedVisitor(this.formatter, this);
    
    this.tree = Array.isArray(tree) ? tree : [tree];
  }

  static deparse(query: Node | Node[], opts: DeparserOptions = {}): string {
    return new Deparser(query, opts).deparseQuery();
  }

  deparseQuery(): string {
    return this.tree
      .map(node => this.deparse(node))
      .join(this.formatter.newline() + this.formatter.newline());
  }

  deparse(node: Node, context: DeparserContext = {}): string | null {
    if (node == null) {
      return null;
    }

    if (typeof node === 'number' || node instanceof Number) {
      return node.toString();
    }

    try {
      return this.consolidatedVisitor.visit(node, context);
    } catch (error) {
      const nodeType = Object.keys(node)[0];
      throw new Error(`Error deparsing ${nodeType}: ${(error as Error).message}`);
    }
  }


}
