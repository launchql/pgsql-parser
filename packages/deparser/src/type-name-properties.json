{
  "TypeCast": ["typeName"],
  "RangeTableFuncCol": ["typeName"],
  "ColumnDef": ["typeName"],
  "XmlSerialize": ["typeName"],
  "JsonOutput": ["typeName"],
  "JsonTableColumn": ["typeName"],
  "CreateDomainStmt": ["typeName"],
  "CreateTransformStmt": ["type_name"],
  "CreateStmt": ["ofTypename"],
  "CreateOpClassItem": ["datatype", "storedtype"],
  "CreateFunctionStmt": ["returnType"],
  "FunctionParameter": ["argType"],
  "CreateCastStmt": ["sourcetype", "targettype"]
} 