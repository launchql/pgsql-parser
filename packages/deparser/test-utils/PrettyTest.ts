import { expectParseDeparse } from './index';

const generateCoded = require('../../../__fixtures__/generated/generated.json');

export class PrettyTest {
  private testCases: string[];
  private testPrefix: string;

  constructor(testCases: string[], testPrefix: string = 'test') {
    this.testCases = testCases;
    this.testPrefix = testPrefix;
  }

  /**
   * Generate individual tests for each test case with both pretty and non-pretty formatting
   */
  generateTests(): void {
    this.testCases.forEach((key, index) => {
      const sql = generateCoded[key];
      const testName = `${this.testPrefix}-${index + 1}`;

      it(`should format ${testName}: ${key} (pretty)`, async () => {
        const result = await expectParseDeparse(sql, { pretty: true });
        expect(result).toMatchSnapshot();
      });

      it(`should format ${testName}: ${key} (non-pretty)`, async () => {
        const result = await expectParseDeparse(sql, { pretty: false });
        expect(result).toMatchSnapshot();
      });
    });
  }

  /**
   * Generate tests with custom test name extraction
   */
  generateTestsWithCustomNaming(nameExtractor: (key: string, index: number) => string): void {
    this.testCases.forEach((key, index) => {
      const sql = generateCoded[key];
      const testName = nameExtractor(key, index);

      it(`should format ${testName}: ${key} (pretty)`, async () => {
        const result = await expectParseDeparse(sql, { pretty: true });
        expect(result).toMatchSnapshot();
      });

      it(`should format ${testName}: ${key} (non-pretty)`, async () => {
        const result = await expectParseDeparse(sql, { pretty: false });
        expect(result).toMatchSnapshot();
      });
    });
  }

  /**
   * Generate a single AST validation test for all test cases
   */
  generateASTValidationTest(): void {
    it('should validate AST equivalence for all test cases', async () => {
      const allSql = this.testCases.map((key) => generateCoded[key]);
      
      for (const sql of allSql) {
        await expectParseDeparse(sql, { pretty: false });
        await expectParseDeparse(sql, { pretty: true });
      }
    });
  }
} 