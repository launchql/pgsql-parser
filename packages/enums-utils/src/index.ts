import toInt from './enums2int.json';
import toStr from './enums2str.json';
import nodes from './nodes.json';

export { toInt, toStr, nodes };
export const getEnum = (enumType, key) => {
  if (typeof key === 'number') {
    return toStr[enumType][key];
  } else {
    return toInt[enumType][key];
  }
};

export * from './types';
