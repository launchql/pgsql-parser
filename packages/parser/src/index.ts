import { Deparser, deparse } from 'pgsql-deparser';
import {
  parseQuery,
  parseQuerySync,
  parsePlPgSQLSync as parseFunction
} from 'libpg-query';

function mapStmt({ stmt, stmt_len, stmt_location }) {
  return {
    RawStmt: {
      stmt,
      stmt_len,
      stmt_location: stmt_location || 0
    }
  };
}

export const parse = (sql) => {
  if (!sql) throw new Error('no SQL provided to parser');
  const result = parseQuerySync(sql);
  return result.stmts.map(mapStmt);
};

export const parseAsync = async (sql) => {
  if (!sql) throw new Error('no SQL provided to parser');
  const result = await parseQuery(sql);
  return result.stmts.map(mapStmt);
};

export { deparse, Deparser, parseFunction };
