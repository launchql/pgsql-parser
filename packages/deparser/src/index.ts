import { Deparser } from "./deparser";

const deparse = Deparser.deparse;
export { deparse, Deparser };
