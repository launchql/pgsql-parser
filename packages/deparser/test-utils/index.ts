import { parse } from '@pgsql/parser';
import { deparse } from '../src';
import { cleanTree, cleanLines } from '../src/utils';
import { readFileSync } from 'fs';
import * as path from 'path';
import { expect } from '@jest/globals';
import {diff} from 'jest-diff'

type ParseErrorType = 
  | 'PARSE_FAILED'
  | 'INVALID_STATEMENT'
  | 'REPARSE_FAILED'
  | 'AST_MISMATCH'
  | 'UNEXPECTED_ERROR'
  | 'INVALID_DEPARSED_SQL';

interface ParseError extends Error {
  type: ParseErrorType;
  relativePath: string;
  sql: string;
  deparsedSql?: string;
  originalAst?: any;
  reparsedAst?: any;
  parseError?: string;
}

function createParseError(
  type: ParseErrorType,
  relativePath: string,
  sql: string,
  deparsedSql?: string,
  originalAst?: any,
  reparsedAst?: any,
  parseError?: string
): ParseError {
  const error = new Error(getErrorMessage(type)) as ParseError;
  error.type = type;
  error.relativePath = relativePath;
  error.sql = sql;
  error.deparsedSql = deparsedSql;
  error.originalAst = originalAst;
  error.reparsedAst = reparsedAst;
  error.parseError = parseError;
  return error;
}

function getErrorMessage(type: ParseErrorType): string {
  switch (type) {
    case 'PARSE_FAILED':
      return 'Parse failed - no statements returned';
    case 'INVALID_STATEMENT':
      return 'Invalid statement structure';
    case 'REPARSE_FAILED':
      return 'Reparse failed - no statements returned';
    case 'AST_MISMATCH':
      return 'AST mismatch after parse/deparse cycle';
    case 'UNEXPECTED_ERROR':
      return 'Unexpected error during parse/deparse cycle';
    case 'INVALID_DEPARSED_SQL':
      return 'Invalid deparsed SQL';
  }
}

export class TestUtils {
  protected printErrorMessage(sql: string, position: number) {
    const lineNumber = sql.slice(0, position).match(/\n/g)?.length || 0;
    const lines = sql.split('\n');
    let colNumber = position - 1;
    for (let l = 0; l < lineNumber; l++) {
      colNumber -= lines[l].length + 1;
    }
    const errMessage = [`Error line ${lineNumber + 1}, column ${colNumber + 1}`];
    if (lineNumber > 0) {
      errMessage.push(lines[lineNumber - 1]);
    }
    errMessage.push(lines[lineNumber]);
    errMessage.push(' '.repeat(colNumber) + '^');
    if (lineNumber < lines.length - 1) {
      errMessage.push(lines[lineNumber + 1]);
    }
    console.error(errMessage.join('\n'));
  }

  tryParse(sql: string) {
    try {
      return parse(sql);
    } catch (err: any) {
      if (err.cursorPosition) {
        this.printErrorMessage(sql, err.cursorPosition);
      }
      throw err;
    }
  }

  expectAstMatch(relativePath: string, sql: string) {
    let tree: any;
    try {
      tree = this.tryParse(sql);
      if (tree.stmts) {
        tree.stmts.forEach((stmt: any) => {
          if (stmt.stmt) {
            const outSql = deparse(stmt.stmt);
            let reparsed;
            try {
              reparsed = parse(outSql);
            } catch (parseErr) {
              throw createParseError(
                'INVALID_DEPARSED_SQL',
                relativePath,
                sql,
                outSql,
                cleanTree([stmt]),
                undefined,
                parseErr instanceof Error ? parseErr.message : String(parseErr)
              );
            }
            const originalClean = cleanTree([stmt]);
            const reparsedClean = cleanTree(reparsed.stmts || []);
            
            if (!tree.stmts) {
              throw createParseError('PARSE_FAILED', relativePath, sql);
            }
            
            if (!stmt.stmt) {
              throw createParseError('INVALID_STATEMENT', relativePath, sql, undefined, cleanTree(tree));
            }
            
            if (!reparsed.stmts) {
              throw createParseError('REPARSE_FAILED', relativePath, sql, outSql, originalClean);
            }
            
            try {
              expect(reparsedClean).toEqual(originalClean);
            } catch (err) {
              throw createParseError('AST_MISMATCH', relativePath, sql, outSql, originalClean, reparsedClean);
            }
          }
        });
      }
    } catch (err) {
      const errorMessages: string[] = [];
      
      if (err instanceof Error && 'type' in err) {
        const parseError = err as ParseError;
        errorMessages.push(`\n❌ ${parseError.type}: ${parseError.relativePath}`);
        errorMessages.push(`❌ INPUT SQL: ${parseError.sql}`);
        
        if (parseError.deparsedSql) {
          errorMessages.push(`❌ DEPARSED SQL: ${parseError.deparsedSql}`);
        }
        
        if (parseError.type === 'INVALID_DEPARSED_SQL') {
          errorMessages.push(
            `\n❌ DEPARSER GENERATED INVALID SQL:`,
            `ORIGINAL AST:`,
            JSON.stringify(parseError.originalAst, null, 2),
            `\nPARSE ERROR: ${parseError.parseError}`
          );
        } else if (parseError.type === 'AST_MISMATCH') {
          errorMessages.push(
            `\n❌ AST COMPARISON:`,
            `EXPECTED AST:`,
            JSON.stringify(parseError.originalAst, null, 2),
            `\nACTUAL AST:`,
            JSON.stringify(parseError.reparsedAst, null, 2),
            `\nDIFF (what's missing from actual vs expected):`,
            diff(parseError.originalAst, parseError.reparsedAst)
          );
        } else if (parseError.originalAst) {
          errorMessages.push(`❌ AST: ${JSON.stringify(parseError.originalAst, null, 2)}`);
        }
      } else {
        // Handle unexpected errors
        errorMessages.push(
          `\n❌ UNEXPECTED ERROR: ${relativePath}`,
          `❌ INPUT SQL: ${sql}`,
          `❌ ERROR: ${err instanceof Error ? err.message : String(err)}`
        );
        if (tree) {
          errorMessages.push(`❌ PARSED AST: ${JSON.stringify(cleanTree(tree), null, 2)}`);
        }
      }
      
      console.log(errorMessages.join('\n'));
      throw err;
    }
  }
}

export class FixtureTestUtils extends TestUtils {
  private fixtures: Record<string, string>;

  constructor() {
    super();
    const GENERATED_JSON = path.join(__dirname, '../../../__fixtures__/generated/generated.json');
    this.fixtures = JSON.parse(readFileSync(GENERATED_JSON, 'utf-8'));
  }

  getTestEntries(filters: string[]) {
    if (filters.length === 0) {
      return Object.entries(this.fixtures);
    }
    
    return Object.entries(this.fixtures).filter(([relPath]) => 
      filters.includes(relPath)
    );
  }

  runFixtureTests(filters: string[]) {
    this.getTestEntries(filters).forEach(([relativePath, sql]) => {
      try {
        this.expectAstMatch(relativePath, sql);
      } catch (err) {
        throw err;
      }
    });
  }
}
