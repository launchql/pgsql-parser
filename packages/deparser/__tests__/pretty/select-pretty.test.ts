import { deparseSync } from '../../src';
import { parse } from 'libpg-query';
import { TestUtils } from '../../test-utils';

describe('Pretty SELECT formatting', () => {
  const basicSelectSql = `SELECT id, name, email FROM users WHERE active = true;`;
  
  const complexSelectSql = `SELECT u.id, u.name, u.email, p.title FROM users u JOIN profiles p ON u.id = p.user_id WHERE u.active = true AND u.created_at > '2023-01-01' GROUP BY u.id, u.name, u.email, p.title HAVING COUNT(*) > 1 ORDER BY u.created_at DESC, u.name ASC LIMIT 10 OFFSET 5;`;

  const selectWithSubquerySql = `SELECT id, name FROM users WHERE id IN (SELECT user_id FROM orders WHERE total > 100);`;

  const selectUnionSql = `SELECT name FROM customers UNION ALL SELECT name FROM suppliers ORDER BY name;`;

  it('should format basic SELECT with pretty option enabled', async () => {
    const parsed = await parse(basicSelectSql);
    const result = deparseSync(parsed, { pretty: true });
    expect(result).toMatchSnapshot();
  });

  it('should maintain single-line format when pretty option disabled', async () => {
    const parsed = await parse(basicSelectSql);
    const result = deparseSync(parsed, { pretty: false });
    expect(result).toMatchSnapshot();
  });

  it('should format complex SELECT with pretty option enabled', async () => {
    const parsed = await parse(complexSelectSql);
    const result = deparseSync(parsed, { pretty: true });
    expect(result).toMatchSnapshot();
  });

  it('should maintain single-line format for complex SELECT when pretty disabled', async () => {
    const parsed = await parse(complexSelectSql);
    const result = deparseSync(parsed, { pretty: false });
    expect(result).toMatchSnapshot();
  });

  it('should format SELECT with subquery with pretty option enabled', async () => {
    const parsed = await parse(selectWithSubquerySql);
    const result = deparseSync(parsed, { pretty: true });
    expect(result).toMatchSnapshot();
  });

  it('should format SELECT with UNION with pretty option enabled', async () => {
    const parsed = await parse(selectUnionSql);
    const result = deparseSync(parsed, { pretty: true });
    expect(result).toMatchSnapshot();
  });

  it('should use custom newline and tab characters in pretty mode', async () => {
    const parsed = await parse(basicSelectSql);
    const result = deparseSync(parsed, { 
      pretty: true, 
      newline: '\r\n', 
      tab: '    ' 
    });
    expect(result).toMatchSnapshot();
  });

  it('should validate AST equivalence between original and pretty-formatted SQL', async () => {
    const testUtils = new TestUtils();
    const testCases = [
      { name: 'basic SELECT', sql: basicSelectSql },
      { name: 'complex SELECT', sql: complexSelectSql },
      { name: 'SELECT with subquery', sql: selectWithSubquerySql },
      { name: 'SELECT with UNION', sql: selectUnionSql }
    ];

    for (const testCase of testCases) {
      const originalParsed = await parse(testCase.sql);
      const prettyFormatted = deparseSync(originalParsed, { pretty: true });
      await testUtils.expectAstMatch(`pretty-${testCase.name}`, prettyFormatted);
    }
  });
});
