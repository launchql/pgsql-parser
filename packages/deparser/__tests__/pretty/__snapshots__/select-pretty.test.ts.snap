// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pretty SELECT formatting should format SELECT with UNION with pretty option enabled 1`] = `
"SELECT
  name
FROM
customers
UNION
ALL
SELECT
  name
FROM
suppliers
ORDER BY
  name;"
`;

exports[`Pretty SELECT formatting should format SELECT with multiple JOINs with pretty option enabled 1`] = `
"SELECT
  u.id,
  u.name,
  u.email,
  p.title
FROM
users AS u 
JOIN profiles AS p ON u.id = p.user_id 
LEFT JOIN orders AS o ON u.id = o.user_id 
RIGHT JOIN addresses AS a ON u.id = a.user_id
WHERE
  u.active = true;"
`;

exports[`Pretty SELECT formatting should format SELECT with subquery with pretty option enabled 1`] = `
"SELECT
  id,
  name
FROM
users
WHERE
  id IN (SELECT
    user_id
  FROM
  orders
  WHERE
    total > 100);"
`;

exports[`Pretty SELECT formatting should format basic SELECT with pretty option enabled 1`] = `
"SELECT
  id,
  name,
  email
FROM
users
WHERE
  active = true;"
`;

exports[`Pretty SELECT formatting should format complex SELECT with pretty option enabled 1`] = `
"SELECT
  u.id,
  u.name,
  u.email,
  p.title
FROM
users AS u 
JOIN profiles AS p ON u.id = p.user_id
WHERE
  u.active = true AND u.created_at > '2023-01-01'
GROUP BY
  u.id,
  u.name,
  u.email,
  p.title
HAVING
  count(*) > 1
ORDER BY
  u.created_at DESC,
  u.name ASC
LIMIT
10
OFFSET
5;"
`;

exports[`Pretty SELECT formatting should maintain single-line format for complex SELECT when pretty disabled 1`] = `"SELECT u.id, u.name, u.email, p.title FROM users AS u JOIN profiles AS p ON u.id = p.user_id WHERE u.active = true AND u.created_at > '2023-01-01' GROUP BY u.id, u.name, u.email, p.title HAVING count(*) > 1 ORDER BY u.created_at DESC, u.name ASC LIMIT 10 OFFSET 5;"`;

exports[`Pretty SELECT formatting should maintain single-line format when pretty option disabled 1`] = `"SELECT id, name, email FROM users WHERE active = true;"`;

exports[`Pretty SELECT formatting should use custom newline and tab characters in pretty mode 1`] = `
"SELECT
    id,
    name,
    email
FROM
users
WHERE
    active = true;"
`;
